`TestGetHeadlessAuthentication` flakiness

## Failure

#### Link(s) to logs

- https://github.com/gravitational/teleport/actions/runs/14889061716/job/41816328966#step:6:2190

#### Relevant snippet

```
    auth_with_roles_test.go:8104: 
        	Error Trace:	/__w/teleport/teleport/lib/auth/auth_with_roles_test.go:8104
        	            				/__w/teleport/teleport/lib/auth/auth_with_roles_test.go:8158
        	Error:      	Error "rpc error: code = DeadlineExceeded desc = stream terminated by RST_STREAM with error code: CANCEL" does not contain "context deadline exceeded"
        	Test:       	TestGetHeadlessAuthentication/NOK_different_user
        	Messages:   	expected context deadline error but got: rpc error: code = DeadlineExceeded desc = stream terminated by RST_STREAM with error code: CANCEL
```

@GavinFrazar this failed in last night's flake detector. 

https://github.com/gravitational/teleport.e/actions/runs/15110062139/job/42467333981

```
=== CONT  TestGetHeadlessAuthentication/NOK_not_found
    auth_with_roles_test.go:8201: 
        	Error Trace:	/__w/teleport.e/teleport.e/lib/auth/auth_with_roles_test.go:8140
        	            				/__w/teleport.e/teleport.e/lib/auth/auth_with_roles_test.go:8201
        	Error:      	Error "rpc error: code = DeadlineExceeded desc = stream terminated by RST_STREAM with error code: CANCEL" does not contain "context deadline exceeded"
        	Test:       	TestGetHeadlessAuthentication/NOK_not_found
        	Messages:   	expected context deadline error but got: rpc error: code = DeadlineExceeded desc = stream terminated by RST_STREAM with error code: CANCEL
--- FAIL: TestGetHeadlessAuthentication/NOK_not_found (1.01s)
```

--- a/api/go.mod
+++ b/api/go.mod
@@ -29,7 +29,7 @@ require (
 	golang.org/x/net v0.39.0
 	golang.org/x/term v0.31.0
 	google.golang.org/genproto/googleapis/rpc v0.0.0-20250428153025-10db94c68c34
-	google.golang.org/grpc v1.72.0
+	google.golang.org/grpc v1.72.1
 	google.golang.org/protobuf v1.36.6
 	gopkg.in/yaml.v2 v2.4.0
 )

--- a/go.mod
+++ b/go.mod
@@ -234,7 +234,7 @@ require (
 	golang.zx2c4.com/wireguard/windows v0.5.3
 	google.golang.org/api v0.230.0
 	google.golang.org/genproto/googleapis/rpc v0.0.0-20250428153025-10db94c68c34
-	google.golang.org/grpc v1.72.0
+	google.golang.org/grpc v1.72.1
 	google.golang.org/grpc/cmd/protoc-gen-go-grpc v1.5.1
 	google.golang.org/protobuf v1.36.6
 	gopkg.in/check.v1 v1.0.0-20201130134442-10cb98267c6c

--- a/integrations/event-handler/go.mod
+++ b/integrations/event-handler/go.mod
@@ -349,7 +349,7 @@ require (
 	google.golang.org/genproto v0.0.0-20250303144028-a0af3efb3deb // indirect
 	google.golang.org/genproto/googleapis/api v0.0.0-20250428153025-10db94c68c34 // indirect
 	google.golang.org/genproto/googleapis/rpc v0.0.0-20250428153025-10db94c68c34 // indirect
-	google.golang.org/grpc v1.72.0 // indirect
+	google.golang.org/grpc v1.72.1 // indirect
 	google.golang.org/grpc/cmd/protoc-gen-go-grpc v1.5.1 // indirect
 	gopkg.in/check.v1 v1.0.0-20201130134442-10cb98267c6c // indirect
 	gopkg.in/evanphx/json-patch.v4 v4.12.0 // indirect

--- a/integrations/terraform/go.mod
+++ b/integrations/terraform/go.mod
@@ -16,7 +16,7 @@ require (
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.10.1
 	github.com/jonboulle/clockwork v0.5.0
 	github.com/stretchr/testify v1.10.0
-	google.golang.org/grpc v1.72.0
+	google.golang.org/grpc v1.72.1
 	google.golang.org/protobuf v1.36.6
 )
 

--- a/lib/auth/auth_with_roles_test.go
+++ b/lib/auth/auth_with_roles_test.go
@@ -40,6 +40,8 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"golang.org/x/crypto/ssh"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 	"google.golang.org/protobuf/types/known/durationpb"
 
 	"github.com/gravitational/teleport"
@@ -8023,7 +8025,13 @@ func TestGetHeadlessAuthentication(t *testing.T) {
 	assertTimeout := func(t require.TestingT, err error, _ ...any) {
 		t.(*testing.T).Helper()
 		require.Error(t, err)
-		require.ErrorContains(t, err, context.DeadlineExceeded.Error(), "expected context deadline error but got: %v", err)
+		s, ok := status.FromError(err)
+		require.True(t, ok)
+		if s.Code() == codes.Unknown {
+			require.ErrorContains(t, err, context.DeadlineExceeded.Error())
+			return
+		}
+		require.Equal(t, codes.DeadlineExceeded.String(), s.Code().String())
 	}
 
 	assertAccessDenied := func(t require.TestingT, err error, _ ...any) {
@@ -8078,12 +8086,6 @@ func TestGetHeadlessAuthentication(t *testing.T) {
 			}
 
 			retrievedHeadlessAuthn, err := client.GetHeadlessAuthentication(ctx, tc.headlessID)
-			// handle context canceled error ambiguity
-			// TODO(gavin): remove this after this issue is fixed:
-			// https://github.com/grpc/grpc-go/issues/8281
-			if err != nil && ctx.Err() != nil {
-				err = trace.Wrap(err, ctx.Err())
-			}
 			tc.assertError(t, err)
 			if err == nil {
 				require.Equal(t, headlessAuthn, retrievedHeadlessAuthn)

